src/cattr/gen.py

```diff
@@ -92,9 +95,31 @@ def make_dict_unstructure_fn(cl, converter, omit_if_default=False, **kwargs):
 
 def make_dict_structure_fn(cl: Type, converter, **kwargs):
     """Generate a specialized dict structuring function for an attrs class."""
+
+    mapping = kwargs.get("mapping", None)
+    if is_generic(cl):
+        base = get_origin(cl)
+        mapping = converter._generate_mapping(cl, mapping)
+        cl = base
+
+    for base in getattr(cl, "__orig_bases__", ()):
+        if is_generic(base) and not str(base).startswith("typing.Generic"):
+            mapping = converter._generate_mapping(base, mapping)
+            break
+
+    if isinstance(cl, TypeVar):
+        cl = getattr(mapping, cl.__name__, cl)
+
     cl_name = cl.__name__
     fn_name = "structure_" + cl_name
-    globs = {"__c_s": converter.structure, "__cl": cl}
+    for p in getattr(cl, "__parameters__", ()):
+        # This is nasty, I am not sure how best to handle `typing.List[str]` or `TClass[int, int]` as a parameter type here
+        name_base = getattr(mapping, p.__name__)
+        name = getattr(name_base, "__name__", str(name_base))
+        name = re.sub("[\[\.\] ,]", "_", name)
```

Comment:
This line raises a Flake8 warning on my end.

Reply:
this should be ripped out and replaced with something nicer unless this way of generating a name is ok and I can replace it.